File: 백준/Gold/9466. 텀 프로젝트/텀 프로젝트.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

var (
	cnt     = 0
	finish  = []bool{}
	visited = []bool{}
	reader  = bufio.NewReader(os.Stdin)
)

func dfs(v, i int, persons []int) {
	visited[v] = true
	var next = persons[v]
	if !visited[next] {
		dfs(next, i, persons)
	} else {
		if !finish[next] {
			i = next
			for i != v {
				cnt += 1
				i = persons[i]
			}
			cnt += 1
		}
	}
	finish[v] = true
}
func search(n int) {
	var persons = make([]int, n+1)
	visited = make([]bool, n+1)
	finish = make([]bool, n+1)
	visited[0] = true
	finish[0] = true
	for j := 1; j <= n; j++ {
		fmt.Fscanf(reader, "%d ", &persons[j])
	}
	for j := 1; j <= n; j++ {
		if !visited[j] {
			dfs(j, j, persons)
		}
	}
	fmt.Println(n - cnt)
}
func main() {

	var tc, n int
	fmt.Fscanf(reader, "%d\n", &tc)
	for i := 0; i < tc; i++ {
		cnt = 0
		fmt.Fscanf(reader, "%d\n", &n)
		search(n)
	}
}

File: 백준/Gold/9466. 텀 프로젝트/텀 프로젝트.py
import sys 
input = sys.stdin.readline
sys.setrecursionlimit(10**6)
tc = int(input())

def dfs(v):
    global cnt, visited, finish
    visited[v]=True
    next = persons[v]
    if not visited[next]:
        dfs(next)
    else:
        if not finish[next]:
            i = next
            while i!=v:
                cnt+=1
                i = persons[i]
            cnt+=1
    finish[v]=True
for _ in range(tc):
    n = int(input())
    persons = [0]+list(map(int, input().split()))
    visited = [True]+[False]*(n)

    finish = [True]+[False]*(n)
    cnt = 0
    for i in range(1, n+1):
        if not visited[i]:
            dfs(i)
    print(n-cnt)
